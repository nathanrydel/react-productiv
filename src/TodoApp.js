import React, { useState } from "react";
import { v4 as uuid } from "uuid";

import TopTodo from "./TopTodo";
import TodoForm from "./TodoForm";
import EditableTodoList from "./EditableTodoList";

/** App for managing a todo list.
 *
 * Props:
 * - initialTodos: possible array of [ todo, ... ]
 *
 * State:
 * - todos: array of [ todo, ... ]
 *
 * App -> TodoApp -> { TodoForm, EditableTodoList }
 */

function TodoApp({ initialTodos }) {
  const [todos, setTodos] = useState(initialTodos);

  console.log('TodoApp renders: ', todos, initialTodos);

  /** add a new todo to list */
  function create(newTodo) {
    const todo = { ...newTodo, id: uuid() };
    setTodos(tasks => [...tasks, todo]);
  }

  /** update a todo with updatedTodo */
  function update(updatedTodo) {
    setTodos(tasks => tasks.map(
      task => task.id === updatedTodo.id ? updatedTodo : task
    ));
  }

  /** delete a todo by id */
  function remove(id) {
    setTodos(tasks => tasks.filter(task => task.id !== id));
  }

  return (
    <main className="TodoApp">
      <div className="row">
        <div className="col-md-6">
        <h3 className="mb-3">Todos</h3>
          {todos.length > 0 &&
            <EditableTodoList
              todos={todos}
              update={update}
              remove={remove}
            />
          }
          {todos.length === 0 &&
            <span
              className="text-muted">
              You have no todos.
            </span>
          }
        </div>

        <div className="col-md-6">
          {todos.length > 0 &&
            <section className="mb-4">
              <h3>Top Todo</h3>
              <TopTodo todos={todos} />
            </section>
          }

          <section>
            <h3 className="mb-3">Add NÃ¼</h3>
            <TodoForm handleSave={create} />
          </section>
        </div>

      </div>
    </main>
  );
}

export default TodoApp;