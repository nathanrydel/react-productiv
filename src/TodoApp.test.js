import React from "react";
import { render } from "@testing-library/react";
import TodoApp from "./TodoApp";

const TODOS = [
  { id: 1, title: "Test1", description: "desc", priority: 1 },
  { id: 2, title: "Test2", description: "desc2", priority: 3 },
  { id: 3, title: "Test3", description: "desc3", priority: 2 },
]

// beforeEach()

// TodoApp smoke Test
describe("TodoApp smoke test", function () {
  test("it renders without crashing", function () {
    render(<TodoApp initialTodos={TODOS} />);
  });
});

// Renders Todos area with 1 Todo Present and Top Todo with a Todo

describe("Rendering TodoApp with content", function () {
  test("Todos area renders with a todo", function () {
    const result = render(<TodoApp initialTodos={TODOS} />)

    expect(result.queryByText("Todos")).toBeInTheDocument();
    expect(result.queryByText("Top Todo")).toBeInTheDocument();
    expect(result.queryAllByText("Test1")).toHaveLength(2);
    expect(result.queryByText("Test2")).toBeInTheDocument();
    expect(result.queryByText("Test3")).toBeInTheDocument();
  })
  // Snapshot of rendered Todos
  test("matches snapshot of correct result", function () {
    const result = render(<TodoApp initialTodos={TODOS} />)

    expect(result).toMatchSnapshot();
  });
});

// Renders Todos area with message You have no todos and TopTodo with No todos yet!
// Snapshot of No Todos

describe("Rendering ToDoApp with no content", function () {
  tests("Todos area without todos", function () {
    const result = render(<TodoApp initialTodos={[]}/>);

    expect(result.queryByText("Todos")).toBeInTheDocument();
    expect(result.queryByText("You have no todos")).toBeInTheDocument();
    expect(result.queryByText("No todos yet!")).toBeInTheDocument();
    expect(result.queryByText("Test1")).not.toBeInTheDocument();
    expect(result.queryByText("Test2")).not.toBeInTheDocument();
    expect(result.queryByText("Test3")).not.toBeInTheDocument();
  });

  // Snapshot of rendered Todos (no todos)
  test("matches snapshot of correct result", function () {
    const result = render(<TodoApp initialTodos={[]} />)

    expect(result).toMatchSnapshot();
  });
});