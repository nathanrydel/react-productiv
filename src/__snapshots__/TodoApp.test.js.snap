// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rendering TodoApp with content matches snapshot of correct result 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <main
        class="TodoApp"
      >
        <div
          class="row"
        >
          <div
            class="col-md-6"
          >
            <h3
              class="mb-3"
            >
              Todos
            </h3>
            <div
              class="EditableTodo"
            >
              <div
                class="mb-3"
              >
                <div
                  class="float-end text-sm-end"
                >
                  <button
                    class="EditableTodo-toggle btn-link btn btn-sm"
                  >
                    Edit
                  </button>
                  <button
                    class="EditableTodo-delBtn btn-link btn btn-sm text-danger"
                  >
                    Del
                  </button>
                </div>
                <div
                  class="Todo"
                  id="1"
                >
                  <div>
                    <b>
                      Test1
                    </b>
                     
                    <small>
                      Priority: 
                      1
                    </small>
                  </div>
                  <div>
                    <small>
                      desc
                    </small>
                  </div>
                </div>
              </div>
            </div>
            <div
              class="EditableTodo"
            >
              <div
                class="mb-3"
              >
                <div
                  class="float-end text-sm-end"
                >
                  <button
                    class="EditableTodo-toggle btn-link btn btn-sm"
                  >
                    Edit
                  </button>
                  <button
                    class="EditableTodo-delBtn btn-link btn btn-sm text-danger"
                  >
                    Del
                  </button>
                </div>
                <div
                  class="Todo"
                  id="2"
                >
                  <div>
                    <b>
                      Test2
                    </b>
                     
                    <small>
                      Priority: 
                      3
                    </small>
                  </div>
                  <div>
                    <small>
                      desc2
                    </small>
                  </div>
                </div>
              </div>
            </div>
            <div
              class="EditableTodo"
            >
              <div
                class="mb-3"
              >
                <div
                  class="float-end text-sm-end"
                >
                  <button
                    class="EditableTodo-toggle btn-link btn btn-sm"
                  >
                    Edit
                  </button>
                  <button
                    class="EditableTodo-delBtn btn-link btn btn-sm text-danger"
                  >
                    Del
                  </button>
                </div>
                <div
                  class="Todo"
                  id="3"
                >
                  <div>
                    <b>
                      Test3
                    </b>
                     
                    <small>
                      Priority: 
                      2
                    </small>
                  </div>
                  <div>
                    <small>
                      desc3
                    </small>
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div
            class="col-md-6"
          >
            <section
              class="mb-4"
            >
              <h3>
                Top Todo
              </h3>
              <div
                class="Todo"
                id="1"
              >
                <div>
                  <b>
                    Test1
                  </b>
                   
                  <small>
                    Priority: 
                    1
                  </small>
                </div>
                <div>
                  <small>
                    desc
                  </small>
                </div>
              </div>
            </section>
            <section>
              <h3
                class="mb-3"
              >
                Add Nü
              </h3>
              <form
                class="NewTodoForm"
              >
                <div
                  class="mb-3"
                >
                  <input
                    aria-label="Title"
                    class="form-control"
                    id="newTodo-title"
                    name="title"
                    placeholder="Title"
                    value=""
                  />
                </div>
                <div
                  class="mb-3"
                >
                  <textarea
                    aria-label="Description"
                    class="form-control"
                    id="newTodo-description"
                    name="description"
                    placeholder="Description"
                  />
                </div>
                <div
                  class="mb-3 d-flex justify-content-between"
                >
                  <div
                    class="w-75 d-flex justify-content-between"
                  >
                    <label
                      class="d-inline-flex"
                      for="newTodo-priority"
                    >
                      Priority:  
                    </label>
                    <select
                      class="form-control form-control-sm d-inline-flex"
                      id="newTodo-priority"
                      name="priority"
                    >
                      <option
                        value="1"
                      >
                        Ultra-Über
                      </option>
                      <option
                        value="2"
                      >
                        Über
                      </option>
                      <option
                        value="3"
                      >
                        Meh
                      </option>
                    </select>
                  </div>
                  <button
                    class="btn-primary rig btn btn-sm NewTodoForm-addBtn"
                  >
                    Gø!
                  </button>
                </div>
              </form>
            </section>
          </div>
        </div>
      </main>
    </div>
  </body>,
  "container": <div>
    <main
      class="TodoApp"
    >
      <div
        class="row"
      >
        <div
          class="col-md-6"
        >
          <h3
            class="mb-3"
          >
            Todos
          </h3>
          <div
            class="EditableTodo"
          >
            <div
              class="mb-3"
            >
              <div
                class="float-end text-sm-end"
              >
                <button
                  class="EditableTodo-toggle btn-link btn btn-sm"
                >
                  Edit
                </button>
                <button
                  class="EditableTodo-delBtn btn-link btn btn-sm text-danger"
                >
                  Del
                </button>
              </div>
              <div
                class="Todo"
                id="1"
              >
                <div>
                  <b>
                    Test1
                  </b>
                   
                  <small>
                    Priority: 
                    1
                  </small>
                </div>
                <div>
                  <small>
                    desc
                  </small>
                </div>
              </div>
            </div>
          </div>
          <div
            class="EditableTodo"
          >
            <div
              class="mb-3"
            >
              <div
                class="float-end text-sm-end"
              >
                <button
                  class="EditableTodo-toggle btn-link btn btn-sm"
                >
                  Edit
                </button>
                <button
                  class="EditableTodo-delBtn btn-link btn btn-sm text-danger"
                >
                  Del
                </button>
              </div>
              <div
                class="Todo"
                id="2"
              >
                <div>
                  <b>
                    Test2
                  </b>
                   
                  <small>
                    Priority: 
                    3
                  </small>
                </div>
                <div>
                  <small>
                    desc2
                  </small>
                </div>
              </div>
            </div>
          </div>
          <div
            class="EditableTodo"
          >
            <div
              class="mb-3"
            >
              <div
                class="float-end text-sm-end"
              >
                <button
                  class="EditableTodo-toggle btn-link btn btn-sm"
                >
                  Edit
                </button>
                <button
                  class="EditableTodo-delBtn btn-link btn btn-sm text-danger"
                >
                  Del
                </button>
              </div>
              <div
                class="Todo"
                id="3"
              >
                <div>
                  <b>
                    Test3
                  </b>
                   
                  <small>
                    Priority: 
                    2
                  </small>
                </div>
                <div>
                  <small>
                    desc3
                  </small>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="col-md-6"
        >
          <section
            class="mb-4"
          >
            <h3>
              Top Todo
            </h3>
            <div
              class="Todo"
              id="1"
            >
              <div>
                <b>
                  Test1
                </b>
                 
                <small>
                  Priority: 
                  1
                </small>
              </div>
              <div>
                <small>
                  desc
                </small>
              </div>
            </div>
          </section>
          <section>
            <h3
              class="mb-3"
            >
              Add Nü
            </h3>
            <form
              class="NewTodoForm"
            >
              <div
                class="mb-3"
              >
                <input
                  aria-label="Title"
                  class="form-control"
                  id="newTodo-title"
                  name="title"
                  placeholder="Title"
                  value=""
                />
              </div>
              <div
                class="mb-3"
              >
                <textarea
                  aria-label="Description"
                  class="form-control"
                  id="newTodo-description"
                  name="description"
                  placeholder="Description"
                />
              </div>
              <div
                class="mb-3 d-flex justify-content-between"
              >
                <div
                  class="w-75 d-flex justify-content-between"
                >
                  <label
                    class="d-inline-flex"
                    for="newTodo-priority"
                  >
                    Priority:  
                  </label>
                  <select
                    class="form-control form-control-sm d-inline-flex"
                    id="newTodo-priority"
                    name="priority"
                  >
                    <option
                      value="1"
                    >
                      Ultra-Über
                    </option>
                    <option
                      value="2"
                    >
                      Über
                    </option>
                    <option
                      value="3"
                    >
                      Meh
                    </option>
                  </select>
                </div>
                <button
                  class="btn-primary rig btn btn-sm NewTodoForm-addBtn"
                >
                  Gø!
                </button>
              </div>
            </form>
          </section>
        </div>
      </div>
    </main>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
